src/main/java/net/oauth/jsontoken/AsyncJsonTokenParser.java

```diff
@@ -0,0 +1,148 @@
+   * This method is not expected to throw exceptions when returning a future. However,
+   * when getting the result of the future, an {@link ExecutionException} may be thrown
+   * in which {@link ExecutionException#getCause()} follows the same possible exceptions
+   * as thrown by {@link JsonTokenParser#verify(JsonToken)}.
+   *
+   * @param jsonToken
+   * @return a {@link ListenableFuture} that will return a Void
+   */
+  public ListenableFuture<Void> verify(JsonToken jsonToken) {
+    ListenableFuture<List<Verifier>> futureVerifiers = provideVerifiers(jsonToken);
+    // Use AsyncFunction instead of Function to allow for checked exceptions to propagate forward
+    AsyncFunction<List<Verifier>, Void> verifyFunction =
+        verifiers -> Futures.immediateFuture(verifyAndReturnVoid(jsonToken, verifiers));
+
+    return Futures.transformAsync(futureVerifiers, verifyFunction, executor);
+  }
+
+  /**
+   * Parses, and verifies, a JSON Token.
+   * This method is not expected to throw exceptions when returning a future. However,
+   * when getting the result of the future, an {@link ExecutionException} may be thrown
+   * in which {@link ExecutionException#getCause()} follows the same possible exceptions
+   * thrown by {@link JsonTokenParser#verifyAndDeserialize(String)}.
+   *
+   * @param tokenString the serialized token that is to parsed and verified.
+   * @return a {@link ListenableFuture} that will return the deserialized {@link JsonObject},
+   * suitable for passing to the constructor of {@link JsonToken}
+   * or equivalent constructor of {@link JsonToken} subclasses.
+   */
+  public ListenableFuture<JsonToken> verifyAndDeserialize(String tokenString) {
+    ListenableFuture<String> futureTokenString = Futures.immediateFuture(tokenString);
+    AsyncFunction<String, JsonToken> verifyFunction =
+        tokenStringInput -> {
+          JsonToken jsonToken = deserialize(tokenStringInput);
+          Function<Void, JsonToken> returnFunction = unused -> jsonToken;
+          return Futures.transform(verify(jsonToken), returnFunction, executor);
+        };
+
+    return Futures.transformAsync(futureTokenString, verifyFunction, executor);
+  }
+
+  /**
+   * Use {@link AsyncVerifierProviders} to get future that will return a list of verifiers
+   * for this token
+   *
+   * @param jsonToken
+   * @return a {@link ListenableFuture} that will return a list of verifiers
+   */
+  private ListenableFuture<List<Verifier>> provideVerifiers(JsonToken jsonToken) {
+    ListenableFuture<JsonToken> futureJsonToken = Futures.immediateFuture(jsonToken);
+    AsyncFunction<JsonToken, List<Verifier>> findVerifiersFunction =
+        token -> asyncVerifierProviders
+            .getVerifierProvider(token.getSignatureAlgorithm())
+            .findVerifier(token.getIssuer(), token.getKeyId());
+
+    Function<List<Verifier>, List<Verifier>> checkNullFunction =
+        verifiers -> {
+          if (verifiers == null) {
```

Comment:
Do we need to check isEmpty too?

Reply:
Verify handles the case that the list of verifiers is empty, so adding it here isn't completely necessary for correct behavior.